crates/server/src/lib.rs

```diff
@@ -217,25 +197,53 @@ impl Server {
             )
         })?;
 
-        let content_base_url = self.config.content_base_url.unwrap_or_else(|| {
-            Url::parse(&format!("http://{addr}", addr = self.config.addr.unwrap())).unwrap()
-        });
-
-        let server = axum::Server::from_tcp(listener)?.serve(
-            create_router(
-                content_base_url,
-                core,
-                temp_dir,
-                files_dir,
-                self.config.content_policy,
-                self.config.record_policy,
-            )
-            .into_make_service(),
+        let content_base_url = self
+            .config
+            .content_base_url
+            .unwrap_or_else(|| Url::parse(&format!("http://{addr}")).unwrap());
+
+        let router = create_router(
+            content_base_url,
+            core,
+            temp_dir,
+            files_dir,
+            self.config.content_policy,
+            self.config.record_policy,
         );
 
-        tracing::info!("listening on {addr}", addr = self.config.addr.unwrap());
+        Ok(InitializedServer {
+            listener,
+            router,
+            core_handle,
+            shutdown: self.config.shutdown,
+        })
+    }
+
+    pub async fn run(self) -> Result<()> {
+        self.initialize().await?.serve().await
+    }
+}
 
-        if let Some(shutdown) = self.config.shutdown {
+pub struct InitializedServer {
```

Comment:
Nit: missing doc comment (aside: perhaps we should slap a `#![deny(missing_docs)]` on this crate since it's intended to be used as a library?).